### PR TITLE
Fix computing available memory on OSX for GC

### LINKED ISSUE #94846: GcMemoryInfo returns incorrect values in .NET 8 on macOS 14 Sonoma
### Description

We build a library that uses the ratio between GCMemoryInfo.MemoryLoadBytes/GCMemoryInfo.TotalAvailableMemoryBytes returned form the GC.GetGCMemoryInfo API to monitor memory consumption an compact a cache when consumption runs above thresholds. We now have a customer report that the cache compact is constantly triggered and it turn out it is because this ratio is seemingly always at 99% when running in .NET 8 on Mac OS 14. Running on .NET 7 on the same machine returns the actual value (corresponding to what is seen in system tools). The value is also correct on Windows and in Linux containers in .NET 8.

### Reproduction Steps

Note: I don't actually have Mac to reproduce this on, so this is a bit hypothetical:

1. Create an app that calls the GC.GetGCMemoryInfo API and outputs/logs the GCMemoryInfo.MemoryLoadBytes/GCMemoryInfo.TotalAvailableMemoryBytes ratio.
2. Add a forced GC.Collect() call before, to make sure the values are updated.
3. Run the app on MacOS 14.

### Expected behavior

The output memory load ratio is roughly the same as can be seen in the OS tools.

### Actual behavior

The output memory load is always basically 100% memory utilization, regardless of the actual memory pressure as can be seen in OS tools.

Note: This only happens on Mac OS X 14, on Windows and Linux (container) the output is the expected.

### Regression?

Yes, in .NET 7 the returned value is correct. 

### Known Workarounds

_No response_

### Configuration

.NET 8 RTM
Mac OS 14
ASP.NET Core application

### Other information

_No response_

### PR DESCRIPTION
**Pull request by janvorli**
We were using the `vm_statistics_data_t::free_count` as the available memory reported to the GC. It turned out this value is only a small portion of the available memory and that the appropriate value should be based on the kern.memorystatus_level value obtained using sysctl. That value represents percentual amount of available memory, so multiplying it by the total memory bytes gets the available memory bytes.

Close #94846

### CODE REVIEW

File: src/coreclr/gc/unix/gcenv.unix.cpp [hunk 1]
```diff
@@ -317,7 +323,24 @@ bool GCToOSInterface::Initialize()
 
     NUMASupportInitialize();
 
-    return true;
+#ifdef TARGET_APPLE
+    const char* mem_free_name = "kern.memorystatus_level";
+    int rc = sysctlnametomib(mem_free_name, NULL, &g_kern_memorystatus_level_mib_length);
+    if (rc == 0)
+    {
+        g_kern_memorystatus_level_mib = (int*)malloc(g_kern_memorystatus_level_mib_length * sizeof(int));
```
Comment (Maoni0) [inline src/coreclr/gc/unix/gcenv.unix.cpp:331]: presumably if this returns null then `sysctlnametomib` would return a failure?

File: src/coreclr/gc/unix/gcenv.unix.cpp [hunk 2]
```diff
@@ -1248,20 +1271,23 @@ uint64_t GetAvailablePhysicalMemory()
...
-    vm_size_t page_size;
-    mach_port_t mach_port;
-    mach_msg_type_number_t count;
-    vm_statistics_data_t vm_stats;
-    mach_port = mach_host_self();
-    count = sizeof(vm_stats) / sizeof(natural_t);
-    if (KERN_SUCCESS == host_page_size(mach_port, &page_size))
+    uint32_t mem_free = 0;
+    size_t mem_free_length = sizeof(uint32_t);
+    if (g_kern_memorystatus_level_mib != NULL)
```
Comment (Maoni0) [inline src/coreclr/gc/unix/gcenv.unix.cpp:1276]: if `GCToOSInterface::Initialize` succeeded, this should always be non null, correct? if so we can just assert.

File: src/coreclr/gc/unix/gcenv.unix.cpp [hunk 3]
```diff
@@ -1248,20 +1271,23 @@ uint64_t GetAvailablePhysicalMemory()
...
     {
-        if (KERN_SUCCESS == host_statistics(mach_port, HOST_VM_INFO, (host_info_t)&vm_stats, &count))
+        int rc = sysctl(g_kern_memorystatus_level_mib, g_kern_memorystatus_level_mib_length, &mem_free, &mem_free_length, NULL, 0);
+        if (rc == 0)
         {
-            available = (int64_t)vm_stats.free_count * (int64_t)page_size;
+            int64_t mem_size = 0;
+            size_t mem_size_length = sizeof(int64_t);
+            int mib[] = { CTL_HW, HW_MEMSIZE };
+            rc = sysctl(mib, 2, &mem_size, &mem_size_length, NULL, 0);
```
Comment (Maoni0) [inline src/coreclr/gc/unix/gcenv.unix.cpp:1284]: do we need to get the total memory every time?

File: src/coreclr/gc/unix/gcenv.unix.cpp (hunk 1 above)
Comment (janvorli) [reply src/coreclr/gc/unix/gcenv.unix.cpp]: Thank you for spotting the missing check, fixed.

**Review by Maoni0 — APPROVED** [review]
LGTM, thanks!

**Pull request merged**